File: examples/governance/model_registry_domain/lib/model_registry_domain/data_usage_agreement/commands/revoke_agreement.rb
module ModelRegistryDomain
  class DataUsageAgreement
    module Commands
      class RevokeAgreement
        include Hecks::Command
        emits "RevokedAgreement"

        attr_reader :agreement_id

        def initialize(agreement_id: nil)
          @agreement_id = agreement_id
        end

        def call
          existing = repository.find(agreement_id)
          if existing
            unless existing.status == "active"
              raise Hecks::Error, "Cannot RevokeAgreement: status must be 'active', got '#{existing.status}'"
            end
            DataUsageAgreement.new(
              id: existing.id,
              model_id: existing.model_id,
              data_source: existing.data_source,
              purpose: existing.purpose,
              consent_type: existing.consent_type,
              effective_date: existing.effective_date,
              expiration_date: existing.expiration_date,
              restrictions: existing.restrictions,
              status: "revoked"
            )
          else
            raise Hecks::Error, "DataUsageAgreement not found: #{agreement_id}"
          end
        end
      end
    end
  end
end

File: examples/governance_static_go/domain/revoke_agreement.go
package domain

import (
	"time"
	"fmt"
)

type RevokeAgreement struct {
	AgreementId string `json:"agreement_id"`
}

func (c RevokeAgreement) CommandName() string { return "RevokeAgreement" }

func (c RevokeAgreement) Execute(repo DataUsageAgreementRepository) (*DataUsageAgreement, *RevokedAgreement, error) {
	existing, err := repo.Find(c.AgreementId)
	if err != nil {
		return nil, nil, err
	}
	if existing == nil {
		return nil, nil, fmt.Errorf("DataUsageAgreement not found: %s", c.AgreementId)
	}
	if existing.Status != "active" {
		return nil, nil, fmt.Errorf("cannot RevokeAgreement: DataUsageAgreement is in %s state", existing.Status)
	}
	existing.Status = "revoked"
	existing.UpdatedAt = time.Now()
	if err := existing.Validate(); err != nil {
		return nil, nil, err
	}
	if err := repo.Save(existing); err != nil {
		return nil, nil, err
	}
	event := RevokedAgreement{
		AggregateID: existing.ID,
		AgreementId: c.AgreementId,
		OccurredAt: time.Now(),
	}
	return existing, &event, nil
}
